File: Solutions/1552. Magnetic Force Between Two Balls/1552. Magnetic Force Between Two Balls.py
class Solution:
    def maxDistance(self, p: List[int], m: int) -> int:
        def foo(pos,dis,m):
            begin,used=-1,0
            for p in pos:
                if(begin<=p):
                    used+=1
                    begin=p+dis
            return used>=m
        p.sort()
        lo,hi=0,p[len(p)-1]-p[0]
        while lo<=hi:
            dis=(lo+hi)//2
            if foo(p,dis,m):
                lo=dis+1
            else:
                hi=dis-1
        return hi

File: Solutions/1552. Magnetic Force Between Two Balls/1552. Magnetic Force Between Two Balls.cpp
class Solution {
public:
    int maxDistance(vector<int>& p, int m) {
        sort(begin(p),end(p));
        int lo=0,hi=p.back()-p.front();
        while(lo<=hi){
            int dis=(lo+hi)/2;
            if(foo(p,dis,m))
                lo=dis+1;
            else{
                hi=dis-1;
            }
        }
        return hi;
    }
    
    bool foo(vector<int>&pos,int dis,int m){
        int used=0,begin=-1;
        for(int p:pos){
            if(begin<=p){
                ++used;
                begin=dis+p;
            }
        }
        return used>=m;
    }
};